redis/src/main/java/org/hibernate/ogm/datastore/redis/RedisDialect.java

```diff
@@ -0,0 +1,546 @@
+import org.hibernate.ogm.datastore.redis.dialect.model.impl.RedisAssociation;
+import org.hibernate.ogm.datastore.redis.dialect.model.impl.RedisAssociationSnapshot;
+import org.hibernate.ogm.datastore.redis.dialect.model.impl.RedisTupleSnapshot;
+import org.hibernate.ogm.datastore.redis.dialect.value.Association;
+import org.hibernate.ogm.datastore.redis.dialect.value.Entity;
+import org.hibernate.ogm.datastore.redis.impl.EntityStorageStrategy;
+import org.hibernate.ogm.datastore.redis.impl.RedisDatastoreProvider;
+import org.hibernate.ogm.datastore.redis.impl.hash.ExperimentalHashEntityStorageStrategy;
+import org.hibernate.ogm.datastore.redis.impl.json.JsonEntityStorageStrategy;
+import org.hibernate.ogm.datastore.redis.impl.json.JsonSerializationStrategy;
+import org.hibernate.ogm.datastore.redis.options.EntityStorageType;
+import org.hibernate.ogm.datastore.redis.options.impl.AssociationStorageOption;
+import org.hibernate.ogm.datastore.redis.options.impl.EntityStorageOption;
+import org.hibernate.ogm.datastore.redis.options.impl.TTLOption;
+import org.hibernate.ogm.dialect.batch.spi.OperationsQueue;
+import org.hibernate.ogm.dialect.spi.AssociationContext;
+import org.hibernate.ogm.dialect.spi.AssociationTypeContext;
+import org.hibernate.ogm.dialect.spi.BaseGridDialect;
+import org.hibernate.ogm.dialect.spi.ModelConsumer;
+import org.hibernate.ogm.dialect.spi.NextValueRequest;
+import org.hibernate.ogm.dialect.spi.TupleContext;
+import org.hibernate.ogm.model.key.spi.AssociationKey;
+import org.hibernate.ogm.model.key.spi.AssociationKeyMetadata;
+import org.hibernate.ogm.model.key.spi.EntityKey;
+import org.hibernate.ogm.model.key.spi.EntityKeyMetadata;
+import org.hibernate.ogm.model.key.spi.RowKey;
+import org.hibernate.ogm.model.spi.AssociationKind;
+import org.hibernate.ogm.model.spi.Tuple;
+import org.hibernate.ogm.options.spi.OptionsContext;
+import org.hibernate.ogm.type.spi.GridType;
+import org.hibernate.persister.entity.Lockable;
+import org.hibernate.type.Type;
+
+import com.lambdaworks.redis.KeyScanCursor;
+import com.lambdaworks.redis.RedisConnection;
+import com.lambdaworks.redis.ScanArgs;
+import com.lambdaworks.redis.protocol.LettuceCharsets;
+
+/**
+ * Stores tuples and associations inside Redis.
+ * <p/>
+ * Tuples are stored in Redis as a JSON serialization of a {@link Entity} object. Associations are stored in Redis obtained as a
+ * JSON serialization of a {@link Association} object either within the entity or external. See {@link RedisProperties#ASSOCIATIONS_STORE} on how to configure
+ * entity or external storage.
+ *
+ * @author Mark Paluch
+ */
+public class RedisDialect extends BaseGridDialect {
+
+	public static final String IDENTIFIERS = "Identifiers";
+	public static final byte[] IDENTIFIERS_AS_BYTES = toBytes( IDENTIFIERS );
+	public static final String ASSOCIATIONS = "Associations";
+
+	protected final Map<EntityStorageType, EntityStorageStrategy> entityStorageStrategies;
```

Comment:
Can you put the `@Immutable` annotation for documentation purposes? Dialects are accessed concurrently, so they must be thread-safe. I've started to use that annotation to document fields save for this despite their declared type not telling this on first sight.